File: MergeIntervals.py
"""
56. Merge Intervals
Given a collection of intervals, merge all overlapping intervals.

For example,
Given [1,3],[2,6],[8,10],[15,18],
return [1,6],[8,10],[15,18].
"""
# O(nlogn)
# O(1)
class Interval(object):
    def __init__(self, s=0, e=0):
        self.start = s
        self.end = e

    def __repr__(self):
        return "[{},{}]".format(self.start,self.end)
        
        
from copy import copy
class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: List[Interval]
        """
        if not intervals:
            return []
        intervals.sort(key = lambda x: x.start)
        res, new = [], copy(intervals[0])
        for i in range(1, len(intervals)):
            if intervals[i].start > new.end:
                res.append(new)
                new = copy(intervals[i])
            else:
                new.end = max(new.end, intervals[i].end)
        res.append(new)
        return res

if __name__ == "__main__":
    intervals = [Interval(1,3), Interval(2,6), Interval(8,10), Interval(15,18)]
    print(Solution().merge(intervals))
    for interval in intervals:
        print(interval, end = ",")

File: Java/MergeIntervals.java
import java.util.*;

public class MergeIntervals {
	
    public List<Interval> merge(List<Interval> intervals) {
        Collections.sort(intervals, new Comparator<Interval>() {
            public int compare(Interval l1, Interval l2) {
                return l1.start - l2.start;
            }
        });
        List<Interval> res = new ArrayList<Interval>();
        if(intervals.size() == 0) {return res;}
        Interval cur = intervals.get(0);
        Interval next;
        for(int i = 1; i < intervals.size(); i++) {
            next = intervals.get(i);
            if(next.start > cur.end) {
                res.add(cur);
                cur = next;
            } else {
                cur.end = Math.max(cur.end, next.end);
            }
        }
        res.add(cur);
        return res;
    }

	public static void main(String[] args) {
		List<Interval> intervals = new ArrayList<Interval>();
		intervals.add(new Interval(2, 6));
		intervals.add(new Interval(15, 18));
		intervals.add(new Interval(8, 10));
		intervals.add(new Interval(1, 3));
		System.out.println(new MergeIntervals().merge(intervals));

	}

}
